### ISSUE TITLE
DOC: Add missing documentation for top-level functions

### ISSUE THREAD

**Issue by eric-wieser** [numpy/numpy#10106]
Exposed are:

* [ ] `<function <lambda> at 0x0000026322836C80>` (`np.deprecate_with_doc`)
* [x] ~`<class 'numpy._import_tools.PackageLoader'>` (`np.PackageLoader`)~ #11703
* [ ] `<class 'numpy.core._internal.TooHardError'>` (`np.TooHardError`)
* [x] ~`<function _numpy_tester at 0x000002632285C268>` (`np._numpy_tester`)~
* [x] `<built-in function compare_chararrays>` (`np.compare_chararrays`) - #12338
* [x] `<built-in function datetime_as_string>` (`np.datetime_as_string`) - #10107 
* [x] `<built-in function datetime_data>` (`np.datetime_data`) -  #10108 
* [x] ~`<built-in function int_asbuffer>` (`np.int_asbuffer`) - #12425~
* [x] `<built-in function nested_iters>` (`np.nested_iters`) - #10009 
* [x] `<function show at 0x00000263225800D0>` (`np.show_config`) - #9258 
* [x] `<class 'numpy.bytes_'>` (`np.bytes0`, `np.string_`, `np.bytes_`)
* [x] `<class 'numpy.character'>` (`np.character`) - #12027
* [x] `<class 'numpy.complexfloating'>` (`np.complexfloating`) - #12027
* [x] `<class 'numpy.datetime64'>` (`np.datetime64`) - #10003, #10039, ...
* [x] `<class 'numpy.flexible'>` (`np.flexible`) - #12027
* [x] `<class 'numpy.float16'>` (`np.half`, `np.float16`) - #11858
* [x] `<class 'numpy.float64'>` (`np.longdouble`, `np.longfloat`) - #11858
* [x] `<class 'numpy.floating'>` (`np.floating`) - #12027
* [x] `<class 'numpy.inexact'>` (`np.inexact`) - #12027
* [x] `<class 'numpy.int32'>` (`np.intc`) - #11858
* [x] `<class 'numpy.integer'>` (`np.integer`) - #12027
* [x] `<class 'numpy.number'>` (`np.number`) - #12027
* [x] `<class 'numpy.signedinteger'>` (`np.signedinteger`) - #12027
* [x] `<class 'numpy.str_'>` (`np.str0`, `np.unicode_`, `np.str_`)
* [x] `<class 'numpy.timedelta64'>` (`np.timedelta64`)
* [x] `<class 'numpy.uint16'>` (`np.uint16`, `np.ushort`) - #11858
* [x] `<class 'numpy.uint32'>` (`np.uintc`) - #11858
* [x] `<class 'numpy.uint32'>` (`np.uint`, `np.uint32`) - #11858
* [x] `<class 'numpy.uint64'>` (`np.uint64`, `np.uintp`, `np.ulonglong`, `np.uint0`) - #11858
* [x] `<class 'numpy.uint8'>` (`np.uint8`, `np.ubyte`) - #11858
* [x] `<class 'numpy.unsignedinteger'>` (`np.unsignedinteger`) - #12027
* [x] `<class 'numpy.void'>` (`np.void0`, `np.void`)

Steps to document many of these:

1. Find the long-form of the name, such as `np.core.multiarray.uint16`
2. Add the documentation ~~to `numpy/add_newdocs.py`~~ i nthe appropriate place
3. Find an appropriate place to link to the function within `:/doc/source/reference/routines.*.rst` - look at existing files to see how

<details>

<summary> Script to generate the above</summary>

```python
from collections.abc import Callable
ndoc = set(
    v
    for k, v in vars(np).items()
    if isinstance(v, Callable) and v.__doc__ is None
)
ndoc = sorted(
    ndoc,
    key=lambda v: (isinstance(v, type) and issubclass(v, np .generic), v.__qualname__)
)
print('\n'.join(
   '* [ ] `{!r}` ({})'.format(
        v,
       ', '.join('`np.{}`'.format(k) for k, v2 in vars(np).items() if v2 is v)
    )
    for v in ndoc
))
```

</details>

**Comment by mohdsanadzakirizvi** [comment]
I would like to work on this as my first contribution, is it open? :)

**Comment by mattip** [comment]
It seems the basic classes are documented, no?

**Comment by eric-wieser** [comment]
`np.float16.__doc__ is None` on my machine. Are you testing with PyPy? I think this might be another case of things happening after `PyType_Ready`.

**Comment by eric-wieser** [comment]
Thanks @YannickJadoul for knocking out a whole bunch of these!

**Comment by smellslikekeenspirit** [comment]
Hi, can I work on the documentation for the unticked functions? Or have they been done already?

**Comment by mattip** [comment]
@smellslikekeenspirit you can check by help(), if it returns None then the documentation is missing. If you find they are **not** missing, please let us know so we can mark them.

**Comment by engineerlisa** [comment]
Hello. Potential first time contributor here. 

I re-ran the script up top and got:
* [ ] `<function <lambda> at 0x0044D300>` (`np.deprecate_with_doc`)
* [ ] `<class 'numpy.TooHardError'>` (`np.TooHardError`)
* [ ] `<function __dir__ at 0x00B1BFA8>` (`np.__dir__`)
* [ ] `<function __getattr__ at 0x009EE780>` (`np.__getattr__`)
* [ ] `<built-in function int_asbuffer>` (`np.int_asbuffer`)
* [ ] `<function loads at 0x04E096F0>` (`np.loads`)
* [ ] `<function show at 0x04D23D20>` (`np.show_config`)
* [ ] `<class 'numpy.bytes_'>` (`np.bytes0`, `np.bytes_`, `np.string_`)
* [ ] `<class 'numpy.datetime64'>` (`np.datetime64`)
* [ ] `<class 'numpy.str_'>` (`np.str0`, `np.unicode_`, `np.str_`)
* [ ] `<class 'numpy.timedelta64'>` (`np.timedelta64`)
* [ ] `<class 'numpy.void'>` (`np.void`, `np.void0`)

I'm struggling to understand what needs to be accomplished. Any guidance would be appreciated.

Looking at np.datetime64 below. I see that np.datetime64.__doc__ is None. But when checking help(), I see output for that. Can someone explain what the current help is returning, and what it is lacking?

```
>>> print(np.datetime64.__doc__)
None
>>> help(np.datetime64)
Help on class datetime64 in module numpy:

class datetime64(generic)
 |  Base class for numpy scalar types.
 |  
 |  Class from which most (all?) numpy scalar types are derived.  For
 |  consistency, exposes the same API as `ndarray`, despite many
 |  consequent attributes being either "get-only," or completely irrelevant.
 |  This is the class from which it is strongly suggested users should derive
 |  custom scalar types.
 |  
 |  Method resolution order:
 |      datetime64
[...]
```

**Comment by rgommers** [comment]
Hi @engineerlisa, welcome! I would suggest to start with `np.show_config`, that's a straightforward function that is quite useful and not yet documented. What you want to do is:
- write a docstring for the function in the same format as other NumPy function have their docstring. (note, look for `generate_config_py` in `numpy/distutils/misc_util.py` for `show()`, that's what generates the whole `__config__.py` file)
- insert an entry for `show_config` in the reference guide. there's not really an obvious place for it, perhaps in https://numpy.org/devdocs/reference/routines.help.html is best. if you see a better place, please suggest that instead

**Comment by eric-wieser** [comment]
Note that `show_config` already has a PR to document it at gh-9312, it would be worth building on the previous attempt.

**Comment by rgommers** [comment]
Hmm, that PR looks a little complicated. @engineerlisa please have a look if that PR makes sense to you (`@ghost` is someone who disappeared from GitHub); the weird formatting stuff in the PR is because the docstring goes through some machinery that generates the `__config__.py` file so simply `""" ... """` content won't work.

@eric-wieser I removed the "good first issue" label, the left overs aren't a matter of writing docs anymore. I think a maintainer should finish the last bits here.

**Comment by engineerlisa** [comment]
Thank you for getting back to me. I understand some parts of that PR. Based on your removing the tag, I'm going to find another candidate for my first contribution.

**Comment by eric-wieser** [comment]
Just two left here

* `np.deprecate_with_doc`, which could do with a docstring similar to `np.deprecate`
* `np.TooHardError`, which I think is raised only by https://numpy.org/doc/stable/reference/generated/numpy.shares_memory.html

Both of these should be easy to document, so I'm going to add back the "good first issue" label

**Comment by kumudlakara** [comment]
Hey, I'm a first time contributor and I would like to take up the remaining two functions if they are still open. Any guidance would be appreciated. Thanks!

**Comment by mattip** [comment]
Hi @kumudlakara. You should start with the [contributor guide](https://numpy.org/doc/stable/dev/index.html) for an overview, especially the section on [documentation](https://numpy.org/doc/stable/dev/index.html#building-docs) and then the [development workflow](https://numpy.org/doc/stable/dev/development_workflow.html) for the details. We don't assign issues, so be sure to mark your PR with `xref gh-10106` to get a cross-reference to show up here, that way people will know you are working on it.

**Comment by kumudlakara** [comment]
Alright, Thank you @mattip .

**Comment by Radhika14soni** [comment]
Hi, I'm a first time contributor and I would like to work on this issue if they are still open. Any guidance would be appreciated. Thanks!

**Comment by mattip** [comment]
Hi @Radhika14soni and welcome. I [commented above](https://github.com/numpy/numpy/issues/10106#issuecomment-731958950) on how to get familiar with the process. Since `deprecate_with_doc` is underway, the only remaining function lacking documentation is `np.TooHardError`. It is raised only by https://numpy.org/doc/stable/reference/generated/numpy.shares_memory.html, which should give you a hint as to the needed docstring.

**Comment by Radhika14soni** [comment]
@mattip Thanks. I'll work on it.

**Comment by Radhika14soni** [comment]
Hello, Am I supposed to make changes in _add_newdocs.py?

**Comment by Radhika14soni** [comment]
@mattip Hello I'd like to work on it. Am I supposed to make changes in _add_newdocs.py?

**Comment by mattip** [comment]
Since this is a pure-python class, the change would be to add a docstring to the [class definition](https://github.com/numpy/numpy/blob/8fee756d8c9d2f5fe211fd9feb999c0da8a89821/numpy/core/_exceptions.py#L119) like for the other exception classes in that file.

**Comment by mugunthanramesh** [comment]
Is there any works pending on this issue that I can take ?

**Comment by melissawm** [comment]
From the comments above it looks like this issue is taken care of, @mugunthanramesh . But feel free to browse other [Documentation-related issues](https://github.com/numpy/numpy/labels/04%20-%20Documentation) and ping me if you need help getting started. Cheers!

**Comment by mattip** [comment]
Closing then. Please open a new issue if there are still missing docstrings, this one is too difficult to navigate.
